// @expo/eas-build-job/dist/archiveSource.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArchiveSourceSchema = exports.ArchiveSourceType = void 0;
const joi_1 = __importDefault(require("@hapi/joi"));
var ArchiveSourceType;
(function (ArchiveSourceType) {
    ArchiveSourceType["S3"] = "S3";
    ArchiveSourceType["URL"] = "URL";
    ArchiveSourceType["PATH"] = "PATH";
})(ArchiveSourceType = exports.ArchiveSourceType || (exports.ArchiveSourceType = {}));
exports.ArchiveSourceSchema = joi_1.default.object({
    type: joi_1.default.string()
        .valid(...Object.values(ArchiveSourceType))
        .required(),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.S3 }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.S3).required(),
        bucketKey: joi_1.default.string().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.URL }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.URL).required(),
        url: joi_1.default.string().uri().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.PATH }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.PATH).required(),
        path: joi_1.default.string().required(),
    }),
});
