// @expo/eas-build-job/dist/ios/baseJob.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.iOSCredentialsSchema = exports.BaseiOSJobSchema = void 0;
const joi_1 = __importDefault(require("@hapi/joi"));
const archiveSource_1 = require("../archiveSource");
const constants_1 = require("../constants");
const iOSCredentialsSchema = joi_1.default.object().keys({
    provisioningProfileBase64: joi_1.default.string().required(),
    distributionCertificate: joi_1.default.object({
        dataBase64: joi_1.default.string().required(),
        password: joi_1.default.string().required(),
    }).required(),
});
exports.iOSCredentialsSchema = iOSCredentialsSchema;
const BaseiOSJobSchema = joi_1.default.object().keys({
    projectArchive: archiveSource_1.ArchiveSourceSchema.required(),
    platform: joi_1.default.string().valid(constants_1.Platform.iOS).required(),
    projectRootDirectory: joi_1.default.string().required(),
    releaseChannel: joi_1.default.string(),
    secrets: joi_1.default.object({
        buildCredentials: iOSCredentialsSchema,
        secretEnvs: joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.string()),
    }).required(),
});
exports.BaseiOSJobSchema = BaseiOSJobSchema;
