// @expo/xdl/build/project/ProjectUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLogger = getLogger;
exports.logWithLevel = logWithLevel;
exports.logDebug = logDebug;
exports.getPlatformTag = getPlatformTag;
exports.logInfo = logInfo;
exports.logError = logError;
exports.logWarning = logWarning;
exports.clearNotification = clearNotification;
exports.attachLoggerStream = attachLoggerStream;

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _Analytics() {
  const data = _interopRequireDefault(require("../Analytics"));

  _Analytics = function () {
    return data;
  };

  return data;
}

function _Logger() {
  const data = _interopRequireDefault(require("../Logger"));

  _Logger = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MAX_MESSAGE_LENGTH = 200;
const _projectRootToLogger = {};

function _getLogger(projectRoot) {
  let logger = _projectRootToLogger[projectRoot];

  if (!logger) {
    logger = _Logger().default.child({
      type: 'project',
      project: _path().default.resolve(projectRoot)
    });
    _projectRootToLogger[projectRoot] = logger;
  }

  return logger;
}

function getLogger(projectRoot) {
  return _getLogger(projectRoot);
}

function logWithLevel(projectRoot, level, fields, msg, id) {
  if (id) {
    fields.issueId = id;
  }

  const logger = _getLogger(projectRoot);

  switch (level) {
    case 'debug':
      logger.debug(fields, msg);
      break;

    case 'info':
      logger.info(fields, msg);
      break;

    case 'warn':
      logger.warn(fields, msg);
      break;

    case 'error':
      logger.error(fields, msg);
      break;

    default:
      logger.debug(fields, msg);
      break;
  }
}

function logDebug(projectRoot, tag, message, id) {
  _getLogger(projectRoot).debug({
    tag
  }, message.toString());
}

function getPlatformTag(platform) {
  const input = platform.toLowerCase().trim();

  switch (input) {
    case 'ios':
      return _chalk().default.bgWhite.black(' iOS ');

    case 'android':
      return _chalk().default.bgGreen.black(' Android ');

    case 'node':
      return _chalk().default.bgCyan.black(' Node ');

    case 'web':
      return _chalk().default.bgMagenta.black(' web ');

    default:
      return _chalk().default.bgWhite.black(` ${platform} `);
  }
}

function logInfo(projectRoot, tag, message, id) {
  const fields = {
    tag
  };

  if (id) {
    fields.issueId = id;
  }

  _getLogger(projectRoot).info(fields, message.toString());
}

function logError(projectRoot, tag, message, id) {
  const fields = {
    tag
  };

  if (id) {
    fields.issueId = id;
  }

  _getLogger(projectRoot).error(fields, message.toString());

  let truncatedMessage = message.toString();

  if (truncatedMessage.length > MAX_MESSAGE_LENGTH) {
    truncatedMessage = truncatedMessage.substring(0, MAX_MESSAGE_LENGTH);
  } // temporarily remove sentry until we can trim events
  // send error to Sentry
  // Sentry.logError(message.toString(), {
  //   tags: { tag },
  // });

}

function logWarning(projectRoot, tag, message, id) {
  const fields = {
    tag
  };

  if (id) {
    fields.issueId = id;
  }

  _getLogger(projectRoot).warn(fields, message.toString());

  let truncatedMessage = message.toString();

  if (truncatedMessage.length > MAX_MESSAGE_LENGTH) {
    truncatedMessage = truncatedMessage.substring(0, MAX_MESSAGE_LENGTH);
  }

  _Analytics().default.logEvent('Project Warning', {
    projectRoot,
    tag,
    message: truncatedMessage
  });
}

function clearNotification(projectRoot, id) {
  _getLogger(projectRoot).info({
    tag: 'expo',
    issueCleared: true,
    issueId: id
  }, `No issue with ${id}`);
}

function attachLoggerStream(projectRoot, stream) {
  _getLogger(projectRoot).addStream(stream);
}
