// @expo/eas-build-job/dist/android/baseJob.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseAndroidJobSchema = void 0;
const joi_1 = __importDefault(require("@hapi/joi"));
const archiveSource_1 = require("../archiveSource");
const constants_1 = require("../constants");
const AndroidKeystoreSchema = joi_1.default.object({
    dataBase64: joi_1.default.string().required(),
    keystorePassword: joi_1.default.string().required(),
    keyAlias: joi_1.default.string().required(),
    keyPassword: joi_1.default.string().required(),
});
const BaseAndroidJobSchema = joi_1.default.object({
    projectArchive: archiveSource_1.ArchiveSourceSchema.required(),
    platform: joi_1.default.string().valid(constants_1.Platform.Android).required(),
    projectRootDirectory: joi_1.default.string().required(),
    releaseChannel: joi_1.default.string(),
    secrets: joi_1.default.object({
        buildCredentials: joi_1.default.object({ keystore: AndroidKeystoreSchema.required() }),
        secretEnvs: joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.string()),
    }).required(),
});
exports.BaseAndroidJobSchema = BaseAndroidJobSchema;
