// @expo/eas-build-job/dist/android/managed/job.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManagedBuildType = exports.ManagedAndroidJobSchema = void 0;
const joi_1 = __importDefault(require("@hapi/joi"));
const constants_1 = require("../../constants");
const baseJob_1 = require("../baseJob");
var ManagedBuildType;
(function (ManagedBuildType) {
    ManagedBuildType["APK"] = "apk";
    ManagedBuildType["AppBundle"] = "app-bundle";
    ManagedBuildType["DevelopmentClient"] = "development-client";
})(ManagedBuildType || (ManagedBuildType = {}));
exports.ManagedBuildType = ManagedBuildType;
const ManagedAndroidJobSchema = baseJob_1.BaseAndroidJobSchema.concat(joi_1.default.object().keys({
    type: joi_1.default.string().valid(constants_1.Workflow.Managed),
    buildType: joi_1.default.string()
        .valid(...Object.values(ManagedBuildType))
        .default(ManagedBuildType.AppBundle),
    username: joi_1.default.string(),
}));
exports.ManagedAndroidJobSchema = ManagedAndroidJobSchema;
